Sources/KSCrashRecording/KSCrashLifecycleHandler.mm

```diff
@@ -0,0 +1,308 @@
+ * events before the app it is being installed in. This can lead
+ * to mismatched user perception (foreground when the app is actually
+ * background). That mismatch leads to wrongly categorizing
+ * issues reported by the system.
+ *
+ * To fix this, we need to swizzle app an scene delegates. This allows
+ * us to receive delegate callbacks before anyone else and correctly
+ * record the state of the app before the app has a chance to run any
+ * code in those transitions and possibly cause a reliability event.
+ *
+ *
+ */
+// We need to have UIApplication to do any of this.
+#if KSCRASH_HAS_UIAPPLICATION
+
+#import "KSCrashAppStateTracker+Private.h"
+
+#import <objc/runtime.h>
+#import <map>
+#import <string>
+
+#import <UIKit/UIKit.h>
+
+@interface __KS_CALLING_DELEGATE_TEMPLATE__ : NSObject <UIApplicationDelegate>
+@end
+
+#define UISCENE_AVAILABLE API_AVAILABLE(ios(13.0))
+
+UISCENE_AVAILABLE
+@interface UIScene (__KS_CALLING_DELEGATE_TEMPLATE__)
+- (void)__ks_proxyDelegate;
+@end
+
+@interface UIApplication (__KS_CALLING_DELEGATE_TEMPLATE__)
+- (void)__ks_setDelegate:(id<UIApplicationDelegate>)delegate;
+@end
+
+UISCENE_AVAILABLE
+@interface __KS_CALLING_DELEGATE_TEMPLATE__ () <UISceneDelegate>
+@end
+
+@implementation __KS_CALLING_DELEGATE_TEMPLATE__
+
+static BOOL SwizzleInstanceMethod(Class klass, SEL originalSelector, SEL swizzledSelector)
+{
+    Method original = class_getInstanceMethod(klass, originalSelector);
+    Method swizzled = class_getInstanceMethod(klass, swizzledSelector);
+    
+    if (!original || !swizzled) {
+        return NO;
+    }
+    
+    method_exchangeImplementations(original, swizzled);
+    return YES;
+}
+
+typedef BOOL (*ApplicationDelegate_TwoArgs)(id, SEL, id, id);
+typedef void (*ApplicationDelegate_OneArg)(id, SEL, id);
+
+static std::map<std::string, Method> gMappings = {};
+
+static void __KS_CALLING_DELEGATE__(id self, SEL cmd, id arg)
+{
+    std::string name(sel_getName(cmd));
+    NSLog(@"[MAP] %s", name.c_str());
+    const auto it = gMappings.find(name);
+    if (it != gMappings.end()) {
+        NSLog(@"[MAP:implemented] %s", name.c_str());
+        ApplicationDelegate_OneArg imp = (ApplicationDelegate_OneArg)method_getImplementation(it->second);
+        imp(self, cmd, arg);
+    }
+}
+
+static BOOL __KS_CALLING_DELEGATE__(id self, SEL cmd, id arg1, id arg2)
+{
+    std::string name(sel_getName(cmd));
+    NSLog(@"[MAP] %s", name.c_str());
+    const auto it = gMappings.find(name);
+    if (it != gMappings.end()) {
+        NSLog(@"[MAP:implemented] %s", name.c_str());
+        ApplicationDelegate_TwoArgs imp = (ApplicationDelegate_TwoArgs)method_getImplementation(it->second);
+        return imp(self, cmd, arg1, arg2);
+    }
+    return YES;
+}
+
++ (void)load
+{
+    static BOOL sDontSwizzle = [NSProcessInfo.processInfo.environment[@"KSCRASH_APP_SCENE_DELEGATE_SWIZZLE_DISABLED"] boolValue];
+    if (sDontSwizzle) {
+        return;
+    }
+#if KSCRASH_HAS_UIAPPLICATION
```

Comment:
You check it in the beginning of the file. In case this check is needed, maybe there is no point to run `+load` if the OS does'n support `UIApplication`?